File: front-end/src/app/domain/models/game/action/impl/money/DebtForeclosureAction.ts
import {Player} from "../../../Player";
import {GameActionUtil} from "../../GameActionUtil";
import {MoneyExchangeAction} from "./MoneyExchangeAction";

/**
 * Representation, of the action, of a foreclosure when a player defaults on debt. After this action the debtor needs to
 * be declared bankrupt.
 *
 * @author Hamza el Haouti
 */
export class DebtForeclosureAction extends MoneyExchangeAction {

  public constructor(
    id: number,
    payee: Player,
    performedAt: Date,
    recipient: Player,
    amount: number,
    reasonOfDebt: MoneyExchangeAction
  ) {
    super(DebtForeclosureAction.name, id, payee, performedAt, recipient, amount);
    this._reasonOfDebt = reasonOfDebt;
  }

  private _reasonOfDebt: MoneyExchangeAction;

  public get reasonOfDebt(): MoneyExchangeAction {
    return this._reasonOfDebt;
  }

  public static trueCopy(action: DebtForeclosureAction): DebtForeclosureAction | null {
    return new DebtForeclosureAction(
      action.id,
      Player.trueCopy(action.actor),
      new Date(action.performedAt),
      action.recipient ? Player.trueCopy(action.recipient) : null,
      action.amount,
      <MoneyExchangeAction>GameActionUtil.trueCopy(action.reasonOfDebt)
    );
  }

  public override toJson(): object {
    return Object.assign(
      super.toJson(), {
        "reasonOfDebt": this.reasonOfDebt.toJson(),
      });
  }

  public override toString(): string {
    return `${this.actor.user.username} could not pay ${this.reasonOfDebt.amount}. This resulted in the ` +
      `player's money ($${this.amount}) being foreclosed, and given to: ${this.recipient == null ? "the bank" : this.recipient.user.username}`;
  }

}

File: back-end/src/main/java/nl/hva/backend/domain/models/game/action/impl/money/moneyExchangeImpl/DebtForeclosureAction.java
package nl.hva.backend.domain.models.game.action.impl.money.moneyExchangeImpl;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import nl.hva.backend.domain.models.game.Player;
import nl.hva.backend.domain.models.game.action.impl.money.MoneyExchangeAction;

import javax.persistence.Entity;
import javax.persistence.ManyToOne;

/**
 * Representation, of the action, of a foreclosure when a player defaults on debt. After this action the debtor needs to
 * be declared bankrupt.
 *
 * @author Hamza el Haouti
 */
@Entity
@Getter
@Setter
@NoArgsConstructor
public class DebtForeclosureAction extends MoneyExchangeAction {

    @ManyToOne
    private MoneyExchangeAction reasonOfDebt;

    /**
     * Constructor for the manual creation of an instance, that is intended to be persisted to JPA.
     *
     * @param debtor          The player instance who owes debt.
     * @param creditor        The party who is owed money, either the bank, or a player.
     * @param amountAvailable The amount the debtor still can pay.
     * @param reasonOfDebt    Action, that resulted in the debt.
     */
    public DebtForeclosureAction(
            Player debtor,
            Player creditor,
            double amountAvailable,
            MoneyExchangeAction reasonOfDebt
    ) {
        super(debtor, creditor, amountAvailable);
        this.reasonOfDebt = reasonOfDebt;
    }

}